Source/Tests/LinuxSyscalls/Syscalls.cpp

```diff
@@ -321,26 +350,60 @@ uint64_t ExecveHandler(const char *pathname, char* const* argv, char* const* env
 
     auto OldArgv = argv;
 
-    // Skip filename argument
-    ++OldArgv;
-    while (*OldArgv) {
-      // Append the arguments together
-      ExecveArgs.emplace_back(*OldArgv);
+    // It is valid to provide nullptr first argument.
+    if (*OldArgv) {
+      // Skip filename argument
       ++OldArgv;
+      while (*OldArgv) {
+        // Append the arguments together
+        ExecveArgs.emplace_back(*OldArgv);
+        ++OldArgv;
+      }
+    }
+    else {
+      // Linux kernel will stick an empty argument in to the argv list if none are provided.
+       ExecveArgs.emplace_back(NullString);
     }
 
     // Emplace nullptr at the end to stop
     ExecveArgs.emplace_back(nullptr);
   }
 
-  if (Args) {
-    Result = ::syscall(SYS_execveat, Args->dirfd, "/proc/self/exe",
-                       const_cast<char *const *>(ExecveArgs.data()), envp, Args->flags);
-  }
-  else {
-    Result = execve("/proc/self/exe", const_cast<char *const *>(ExecveArgs.data()), envp);
+  if (IsFDExec) {
+    if (envp) {
+      auto OldEnvp = envp;
+      while (*OldEnvp) {
+        EnvpArgs.emplace_back(*OldEnvp);
+        ++OldEnvp;
+      }
+    }
+
+    int Flags = fcntl(Args.dirfd, F_GETFD);
+    if (Flags & FD_CLOEXEC) {
+      // We need to duplicate the FD if FD_CLOEXEC is set.
+      // FEX needs the FD to live past execve when binfmt_misc isn't used.
+      Args.dirfd = dup(Args.dirfd);
+    }
+
+    // Remove AT_EMPTY_PATH now. Can't exist in this path.
+    Args.flags &= ~AT_EMPTY_PATH;
+
+    // Create the environment variable to pass the FD to our FEX.
+    // Needs to stick around until execveat completes.
+    FDExecEnv = "FEX_EXECVEFD=" + std::to_string(Args.dirfd);
+
+    // Insert the FD for FEX to track.
+    EnvpArgs.emplace_back(FDExecEnv.data());
```

Comment:
Where is this extra argument actually used?

Reply:
It's used in the next FEXInterpreter instance that is getting executed.

Reply:
Where can I see this in the code? Do we actively have to extract it again from envp on the other end, or does this happen indirectly?

Reply:
https://github.com/FEX-Emu/FEX/pull/2334/files#diff-cd1944a762a8c051124ab5532528cddff1933f5c0d2e3a45785c36751868b244R224
We extract it from envp here.

Reply:
Ah, yes - I was somehow looking for a manual `envp` readout as opposed to `getenv`. Makes sense!